### PR TITLE
NET-68: ecdsa signature length

### PR DESCRIPTION
**Pull request by ooki2day**
**- What I did**

Fixed ecdsa signature variable length

### CODE REVIEW

File: network/servicenetwork/servicenetwork_test.go [hunk 1]
```diff
@@ -353,3 +362,22 @@ func TestServiceNetwork_SendCascadeMessage2(t *testing.T) {
 	require.True(t, success)
 }
 */
+
+func TestEcdsaMarshalUnmarshal(t *testing.T) {
```
Comment (dmitry-panchenko) [inline network/servicenetwork/servicenetwork_test.go]: Move this test to platformpolicy package.

File: network/servicenetwork/servicenetwork_test.go [hunk 2]
```diff
@@ -353,3 +362,22 @@ func TestServiceNetwork_SendCascadeMessage2(t *testing.T) {
...
+	count := 1000
+	data := testutils.RandomRef()
+	for i := 0; i < count; i++ {
+		privateKey, err := foundation.GeneratePrivateKey()
```
Comment (dmitry-panchenko) [inline network/servicenetwork/servicenetwork_test.go]: Do not use `foundation`. Instead you should use `core.KeyProcessor` component.

File: platformpolicy/internal/sign/wrapper.go [hunk 3]
```diff
@@ -81,13 +79,41 @@ type ecdsaVerifyWrapper struct {
...
+	ecdsaSignature := ecdsaSignature{r, s}
 	hash := sw.hasher.Hash(data)
 
 	return ecdsa.Verify(sw.publicKey, hash, ecdsaSignature.R, ecdsaSignature.S)
 }
+
+func makeSignature(r, s *big.Int) []byte {
+	if (len(r.Bytes()) > BigIntLength) ||
+		(len(s.Bytes()) > BigIntLength) {
+		panic("[ makeSignature ] wrong r, s length")
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: Write length that caused panic.

File: platformpolicy/internal/sign/wrapper.go [hunk 4]
```diff
@@ -81,13 +79,41 @@ type ecdsaVerifyWrapper struct {
...
+	res := make([]byte, rLen+sLen+2) // 2 bytes for len
+	res[0] = rLen
+	copy(res[1:rLen+2], r.Bytes())
+	res[rLen+1] = sLen
+	copy(res[rLen+2:], s.Bytes())
+	return res[:]
+}
+
+func getRSFromBytes(data []byte) (*big.Int, *big.Int) {
+	if len(data) > (BigIntLength*2 + 2) {
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: What means `+2`?**

File: platformpolicy/internal/sign/wrapper.go [hunk 5]
```diff
@@ -26,14 +26,12 @@ import (
 	"github.com/pkg/errors"
 )
 
+const BigIntLength = 32
```
Comment (AndreyBronin) [inline platformpolicy/internal/sign/wrapper.go]: BigIntLength might be not exported - bigIntLength, coz it uses only in this package

File: network/servicenetwork/servicenetwork_test.go (hunk 1 above)
Comment (ooki2day) [reply network/servicenetwork/servicenetwork_test.go]: done

File: network/servicenetwork/servicenetwork_test.go (hunk 2 above)
Comment (ooki2day) [reply network/servicenetwork/servicenetwork_test.go]: done

File: platformpolicy/internal/sign/wrapper.go (hunk 4 above)
Comment (ooki2day) [reply platformpolicy/internal/sign/wrapper.go]: done

File: platformpolicy/internal/sign/wrapper.go (hunk 5 above)
Comment (ooki2day) [reply platformpolicy/internal/sign/wrapper.go]: done

File: platformpolicy/internal/sign/wrapper.go [hunk 6]
```diff
@@ -81,13 +80,43 @@ type ecdsaVerifyWrapper struct {
...
+	res := make([]byte, rLen+sLen+2) // 2 bytes for len
+	res[0] = rLen
+	copy(res[1:rLen+2], r.Bytes())
+	res[rLen+1] = sLen
+	copy(res[rLen+2:], s.Bytes())
+	return res[:]
+}
+
+func getRSFromBytes(data []byte) (*big.Int, *big.Int) {
+	if len(data) > (bigIntLength*2 + 2) { // 2 bytes for len
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: Move that `2 bytes for len` to constant and use it.

File: platformpolicy/internal/sign/wrapper.go [hunk 7]
```diff
@@ -81,13 +81,43 @@ type ecdsaVerifyWrapper struct {
...
+}
+
+func getRSFromBytes(data []byte) (*big.Int, *big.Int) {
+	if len(data) > (bigIntLength*2 + lenBytes) {
+		err := fmt.Sprintf("[ getRSFromBytes ] wrong data length to get a r, s. recv len: %d", len(data))
+		panic(err)
+	}
+	r := new(big.Int)
+	s := new(big.Int)
+	rLen := data[0]
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: Return error when you got `len(data) == 0`

File: platformpolicy/internal/sign/wrapper.go [hunk 8]
```diff
@@ -81,13 +81,43 @@ type ecdsaVerifyWrapper struct {
...
+	sLen := data[rLen+1]
+	rBytes := make([]byte, rLen)
+	sBytes := make([]byte, sLen)
+	copy(rBytes, data[1:rLen+2])
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: `rLen+2`?

File: platformpolicy/internal/sign/wrapper.go [hunk 9]
```diff
@@ -81,13 +81,43 @@ type ecdsaVerifyWrapper struct {
...
+	copy(sBytes, data[rLen+2:])
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: `rLen+2`?

File: platformpolicy/internal/sign/wrapper.go [hunk 10]
```diff
@@ -81,13 +81,43 @@ type ecdsaVerifyWrapper struct {
...
+	if (len(r.Bytes()) > bigIntLength) ||
+		(len(s.Bytes()) > bigIntLength) {
+		err := fmt.Sprintf("[ makeSignature ] wrong r, s length. r: %d; s: %d; needed: %d", len(r.Bytes()), len(s.Bytes()), bigIntLength)
+		panic(err)
+	}
+	rLen := uint8(len(r.Bytes()))
+	sLen := uint8(len(s.Bytes()))
+	res := make([]byte, rLen+sLen+lenBytes)
+	res[0] = rLen
+	copy(res[1:rLen+2], r.Bytes())
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: `rLen+2`?

File: platformpolicy/internal/sign/wrapper.go [hunk 11]
```diff
@@ -81,13 +81,43 @@ type ecdsaVerifyWrapper struct {
...
+	res[rLen+1] = sLen
+	copy(res[rLen+2:], s.Bytes())
```
Comment (dmitry-panchenko) [inline platformpolicy/internal/sign/wrapper.go]: `rLen+2`?

**Pull request merged**